test/utils/Checkpoints.t.sol

```diff
@@ -0,0 +1,276 @@
+        }
+    }
+
+    function testLookup(uint32[] memory keys, uint224[] memory values, uint32 lookup) public {
+        vm.assume(values.length > 0);
+        _prepareKeys(keys, 64);
+
+        uint32 lastKey = keys.length == 0 ? 0 : keys[keys.length - 1];
+        lookup = _boundUint32(lookup, 0, lastKey + 64);
+
+        uint224 upper = 0;
+        uint224 lower = 0;
+        uint32 lowerKey = type(uint32).max;
+        for (uint256 i = 0; i < keys.length; ++i) {
+            uint32 key = keys[i];
+            uint224 value = values[i % values.length];
+
+            // push
+            _ckpts.push(key, value);
+
+            // track expected result of lookups
+            if (key <= lookup) {
+                upper = value;
+            }
+            if (key >= lookup && (i == 0 || keys[i - 1] < lookup)) {
+                lowerKey = key;
+            }
+            if (key == lowerKey) {
+                lower = value;
+            }
+        }
+
+        // check lookup
+        assertTrue(_ckpts.lowerLookup(lookup) == lower);
+        assertTrue(_ckpts.upperLookup(lookup) == upper);
+        assertTrue(_ckpts.upperLookupRecent(lookup) == upper);
+    }
+}
+
+contract CheckpointsTrace160Test is Test {
+    using Checkpoints for Checkpoints.Trace160;
+
+    Checkpoints.Trace160 internal _ckpts;
+
+    // helpers
+    function _boundUint96(uint96 x, uint96 min, uint96 max) internal view returns (uint96) {
+        return SafeCast.toUint96(bound(uint256(x), uint256(min), uint256(max)));
+    }
+
+    function _prepareKeys(uint96[] memory keys, uint96 maxSpread) internal view {
+        uint96 lastKey = 0;
+        for (uint256 i = 0; i < keys.length; ++i) {
+            uint96 key = _boundUint96(keys[i], lastKey, lastKey + maxSpread);
+            keys[i] = key;
+            lastKey = key;
+        }
+    }
+
+    function _assertLatestCheckpoint(bool exist, uint96 key, uint160 value) internal {
+        (bool _exist, uint96 _key, uint160 _value) = _ckpts.latestCheckpoint();
+        assertTrue(_exist == exist);
+        assertTrue(_key == key);
+        assertTrue(_value == value);
+    }
+
+    // tests
+    function testPush(uint96[] memory keys, uint160[] memory values) public {
+        vm.assume(values.length > 0);
+        _prepareKeys(keys, 64);
+
+        // initial state
+        assertTrue(_ckpts.length() == 0);
+        assertTrue(_ckpts.latest() == 0);
+        _assertLatestCheckpoint(false, 0, 0);
+
+        uint256 duplicates = 0;
+        for (uint256 i = 0; i < keys.length; ++i) {
+            uint96 key = keys[i];
+            uint160 value = values[i % values.length];
+            if (i > 0 && key == keys[i - 1]) ++duplicates;
+
+            // push
+            _ckpts.push(key, value);
+
+            // check length & latest
+            assertTrue(_ckpts.length() == i + 1 - duplicates);
+            assertTrue(_ckpts.latest() == value);
+            _assertLatestCheckpoint(true, key, value);
+        }
+    }
```

Comment:
I have the feeling this function is checking multiple things:

- Initial state
- Duplicates
- `length` and `latest` consistency

I think this is a similar conversation to the one we had for the Certora rules (smaller tests vs a single big one). However, in the case of fuzzing, I see value in making individual tests so we can configure the fuzzer on purpose. For example, not bounding the `keys` may make sense for testing duplicates.

Do you think makes sense splitting the tests?

Reply:
I'm not sure how splitting that test would make it better. I believe it would decrease readability. 

This is about checking that the `push` method updates the structure correctly. We do the same in FV:
- run an arbitrary push function
- check that the affected view functions are updated correctly. (`length()`, `latest()`, `latestCheckpoint()`)

Reply:
It's true that this function is checking for a lot of things but I think it makes sense to keep them coupled in one test. Because we want to test a sequence of pushes and assert properties at each step.

Reply:
> Because we want to test a sequence of pushes and assert properties at each step.

Okay now I see the point, but still, feel the initial assertions are unnecessary. Both keeping them or removing is fine for me so I'm solving

Reply:
That's true, the initial state is kind of unnecessary to check every time. I'd let @Amxx decide whether to move that to a separate test (which wouldn't be fuzzed since it's just the initial state).

Reply:
The first check is also a sanity check, that makes we are starting with a clean state. Given our limited understanding of how foundry setup/checkpoint the contract, I felt it was better to have it.